qmxgraph/common_testing.py

```diff
@@ -1,34 +1,74 @@
 """
 Those are helper tools that could help in applications' tests.
 """
+from contextlib import ExitStack, contextmanager, suppress
+from typing import Iterator, List, cast
 
+from PyQt5.QtCore import pyqtSignal
 
-def get_cell_count(widget, filter_function):
+from qmxgraph.callback_blocker import CallbackBlocker
+from qmxgraph.widget import QmxGraph
+
+
+def get_cell_count(widget: QmxGraph, filter_function: str) -> int:
     """
-    :param qmxgraph.widget.QmxGraph widget: The widget used to display the
-        graph.
-    :param str filter_function: A javascript function used to filter the
-        cells.
-    :rtype: int
+    :param widget: The widget used to display the graph.
+    :param filter_function: A javascript function used to filter the cells.
+        This function will receive one argument of the type mxCell and should
+        return a boolean (`true` select the cell, `false` ignore the cell).
+    :return: The number of selected cells.
     """
     return len(get_cell_ids(widget, filter_function))
 
 
-def get_cell_ids(widget, filter_function):
+def get_cell_ids(widget: QmxGraph, filter_function: str) -> List[str]:
+    """
+    :param widget: The widget used to display the graph.
+    :param filter_function: A javascript function used to filter the cells.
+        This function will receive one argument of the type mxCell and should
+        return a boolean (`true` select the cell, `false` ignore the cell).
+    :return: A list with the id's of the selected cells.
+    """
+    cells_ids = widget.inner_web_view().eval_js(
+        f'''
+        (function(){{
+            var all_cells = api._graphEditor.graph.model.cells;
+            var cells = Object.keys(all_cells).map(
+                function(id){{ return all_cells[id]; }}
+            );
+
+            cells = cells.filter({filter_function});
+            return cells.map(function(aCell){{ return aCell.getId(); }});
+        }})()'''
+    )
+    return cast(List[str], cells_ids)
+
+
+@contextmanager
+def wait_signals(
+    *signals: pyqtSignal,
+    timeout: int = CallbackBlocker.DEFAULT_TIMEOUT,
+) -> Iterator[List[CallbackBlocker]]:
```

Comment:
```suggestion
) -> Generator[List[CallbackBlocker], None, None]:
```

Reply:
I often have difficulty annotating context managers.
before I annotated this I did hit google:

https://stackoverflow.com/a/49736916/783219

> ...
> We can find the stubs for contextlib ...
> ...
> It states that the decorator takes in a `Callable[..., Iterator[_T]]` -- a function with arbitrary arguments returning some iterator. So in conclusion, it would be fine to do:
> ```
> @contextlib.contextmanager
> def foo() -> Iterator[None]:
>     yield
> ```
> So, why does using `Generator[None, None, None]` also work, as suggested by the comments?
> It's because Generator is a subtype of Iterator ...

Reply:
Yeah, I'm not 100% sure of the difference two, but I made the same question to @bluetech (the pytest type-annotation expert) and he said the Generator annotation is more usual for `contextmanager` specifically. 😁

Reply:
OK.

...
but...

```
>>> @contextmanager
... class A:
...   def __init__(self, n):
...     self.n = n
...   def __next__(self):
...     if self.n is None:
...       raise StopIteration()
...     else:
...       n, self.n = self.n, None
...       return n
...
>>> with A('foobar') as x: print(x)
...
foobar
>>> @contextmanager
... def s(n):
...   yield n
...
>>> with s('fizzbuzz') as x: print(x)
...
fizzbuzz
```

The objects bound to `A` and `s` yields the same results but `A` is not an generator.
Also from https://docs.python.org/3/library/typing.html#typing.Generator:

> ... `Generator[YieldType, SendType, ReturnType]`.
> ...
> If your generator will only yield values, set the `SendType` and `ReturnType` to `None`:
> ```
> def infinite_stream(start: int) -> Generator[int, None, None]:
>     while True:
>         yield start
>         start += 1
> ```
> Alternatively, annotate your generator as having a return type of either `Iterable[YieldType]` or `Iterator[YieldType]`:
> ```
> def infinite_stream(start: int) -> Iterator[int]:
>     while True:
>         yield start
>         start += 1
> ```

The function used is a generator but iterator is an accepted decoration.